src/Halogen/Component.purs

```diff
@@ -206,30 +220,33 @@ parentFinalizers
   -> ParentState s s' f f' g p
   -> Array (Finalized g)
 parentFinalizers eval fin (ParentState s) =
-  foldMap childFin s.children <> parentFin
+  foldMap childFin s.children <> foldMap parentFin fin
   where
-  parentFin :: Array (Finalized g)
-  parentFin = case fin of
-    Just i -> [ finalized eval (parentState s.parent) (left i) ]
-    _ -> []
+  parentFin :: f Unit -> Array (Finalized g)
+  parentFin i = [ finalized eval (parentState s.parent) (left i) ]
 
-  childFin :: Tuple (Component s' f' g) s' -> Array (Finalized g)
-  childFin (Tuple (Component c) s) = c.finalizers s
+  childFin :: ChildState s' f f' g p -> Array (Finalized g)
+  childFin child = finalizeComponent child.component child.state
 
 -- | The type used by component containers for their state where `s` is the
 -- | state local to the container, `p` is the type of slot used by the
 -- | container, and the remaining parameters are the type variables for the
 -- | child components.
 newtype ParentState s s' f f' g p = ParentState
   { parent :: s
-  , children :: M.Map p (Tuple (Component s' f' g) s')
-  , memo :: M.Map p (Tree (ParentQuery f f' p) p)
+  , children :: M.Map p (ChildState s' f f' g p)
+  }
+
+type ChildState s' f f' g p =
+  { component :: Component s' f' g
+  , state :: s'
+  , memo :: Maybe (Tree (ParentQuery f f' p) p)
   }
```

Comment:
Ok cool, I wasn't sure if it would help much, just a thought as switching to the pseudo class representation for the data constructors did wonders for PS performance generally, but that was at the "macro scale" I guess!